src/backend/distributed/transaction/backend_data.c

```diff
@@ -1278,3 +1289,119 @@ DecrementExternalClientBackendCounter(void)
 {
 	pg_atomic_sub_fetch_u32(&backendManagementShmemData->externalClientBackendCounter, 1);
 }
+
+
+/*
+ * GetEpochTimeMs returns the epoch value in milliseconds.
+ */
+uint64
+GetEpochTimeMs(void)
+{
+	struct timeval tp;
+
+	gettimeofday(&tp, NULL);
+
+	uint64 result = (uint64) (tp.tv_sec) * 1000;
+	result = result + (uint64) (tp.tv_usec) / 1000;
+	return result;
+}
+
+
+/*
+ * get_cluster_clock() is an UDF that returns a monotonically increasing logical
+ * clock. Clock guarantees to never go back in value after restarts, and makes
+ * best attempt to keep the value close to unix epoch time in milliseconds.
+ */
+Datum
+get_cluster_clock(PG_FUNCTION_ARGS)
+{
+	uint64 epochValue = GetEpochTimeMs();
+	uint64 nextClusterClockValue = 0;
+
+	SpinLockAcquire(&backendManagementShmemData->clockMutex);
+
+	/* Check if the clock is adjusted after the boot */
+	if (backendManagementShmemData->clusterClockValue == 0)
+	{
+		SpinLockRelease(&backendManagementShmemData->clockMutex);
+		ereport(LOG, (errmsg("backend never adjusted the clock, please retry")));
+
+		/* never reaches */
+		PG_RETURN_UINT64(nextClusterClockValue);
+	}
+
+	nextClusterClockValue = backendManagementShmemData->clusterClockValue + 1;
+
+	/*
+	 * Incrementing by 1 ensures uniqueness as the clock precision can be
+	 * less than a millisecond.
+	 */
+	epochValue = epochValue + 1;
+
+	if (epochValue > nextClusterClockValue)
+	{
+		nextClusterClockValue = epochValue;
+	}
+
+	backendManagementShmemData->clusterClockValue = nextClusterClockValue;
+
+	SpinLockRelease(&backendManagementShmemData->clockMutex);
+
+	PG_RETURN_UINT64(nextClusterClockValue);
+}
+
+
+/*
+ * GetClusterClockValue returns the current logical clock value.
+ */
+uint64
+GetClusterClockValue(void)
+{
+	SpinLockAcquire(&backendManagementShmemData->clockMutex);
+	uint64 epochValue = backendManagementShmemData->clusterClockValue;
+	SpinLockRelease(&backendManagementShmemData->clockMutex);
+
+	return epochValue;
+}
+
+
+/*
+ * AdjustAndInitializeClusterClock compares the current epoch value with
+ * the persisted clock value to see if the clock drifted backwards, and
+ * adjusts the logical clock accordingly.
+ * Ideally, this routine should be called only _once_ at the time of boot.
+ */
+void
+AdjustAndInitializeClusterClock(void)
+{
+	uint64 epochPersistedValue;
+
+	if (!GetLocalClockValue(&epochPersistedValue))
+	{
+		/* couldn't get the value from catalog */
```

Comment:
```suggestion
		/*
		 * Couldn't get the value from catalog, might happen
		 * during pg upgrade.
		 */
```

Reply:
Also, should we call GetLocalClockValue after following check, for efficiency ?

```c
	/* Ensure clock never drifts back */
	SpinLockAcquire(&backendManagementShmemData->clockMutex);

	/* Check if the clock was already set after the boot */
	if (backendManagementShmemData->clusterClockValue > 0)
	{
		/* Already initialized */
		SpinLockRelease(&backendManagementShmemData->clockMutex);
		return;
	}

	if (!GetLocalClockValue(&epochPersistedValue))
	{
		/*
		 * Couldn't get the value from catalog, might happen
		 * during pg upgrade.
		 */
....
....
....
```

Reply:
Initially I thought so, but we should never do heavyweight stuff *under spinlock*